components/sup/src/manager/user_config_watcher.rs

```diff
@@ -0,0 +1,395 @@
+    fn name(&self) -> &str {
+        &self.pkg.name
+    }
+
+    fn user_config_path(&self) -> &UserConfigPath {
+        &self.cfg.user_config_path
+    }
+
+    fn service_group(&self) -> &ServiceGroup {
+        &self.service_group
+    }
+}
+
+
+// WorkerState contains the channels the worker uses to communicate
+// with the Watcher.
+struct WorkerState {
+    // This receiver is used by the watcher to be notified when a
+    // worker has events.  The channel is a SyncChannel with buffer
+    // size 1, as we are only interested in the fact that there were
+    // events, not how many there were.
+    have_events: Receiver<()>,
+    // This sender is used by the watcher to notify a worker to stop
+    // running.  It is an async channel because we never want the
+    // UserConfigWatcher to block, even if the receiver end of the
+    // channel somehow dies and/or fails to consume the message.
+    stop_running: Sender<()>,
+    // This receiver is used by the watcher tests to be notified when
+    // a worker finished setting up the watcher and is about to
+    // starting looping it.
+    //
+    // Silence the dead code warnings from rustc, because it is only
+    // used in tests for synchronization purposes.
+    #[allow(dead_code)]
+    started_watching: Receiver<()>,
+}
+
+type ServiceName = String;
+pub struct UserConfigWatcher {
+    states: HashMap<ServiceName, WorkerState>,
+}
+
+impl UserConfigWatcher {
+    pub fn new() -> Self {
+        Self { states: HashMap::new() }
+    }
+
+    /// Adds a service to the User Config Watcher, thereby starting a watcher thread.
+    pub fn add<T: Serviceable>(&mut self, service: &T) -> io::Result<()> {
+        // It isn't possible to use the `or_insert_with` function here because it can't have a
+        // return value, which we need to return the error from `Worker::run`.
+        if self.states.get(service.name()).is_none() {
+            let user_toml_path = match service.user_config_path() {
+                &UserConfigPath::Recommended(ref p) => p.join(USER_CONFIG_FILE),
+                &UserConfigPath::Deprecated(ref p) => {
+                    outputln!(
+                        preamble service.service_group(),
+                        "Not watching {}, because it is located in deprecated path ({}).",
+                        USER_CONFIG_FILE,
+                        p.display(),
+                    );
+                    return Ok(());
+                }
+            };
+            // Establish bi-directional communication with the worker by creating two channels.
+            // The sync_channel's buffer size is 1 because we want to use it as a boolean, i.e. we
+            // are not interested in the events themselves, but only whether at least one has
+            // happened.
+            let (events_tx, events_rx) = sync_channel(1);
+            let (running_tx, running_rx) = channel();
+            let (watching_tx, watching_rx) = sync_channel(1);
```

Comment:
Out of curiosity, why did you choose to go with channels with a buffer size of 1 to act as a boolean, rather than using an `AtomicBool`?

Reply:
Uh, I guess that it has some Golang background, but I'll leave this question to @asymmetric - he wrote that code. ;)

Reply:
@asymmetric any input on this one? ☝️

Reply:
Sorry for the delay, I was on holiday. Mostly because I found easier to reason about this problem using message passing, rather than sharing data.

If I remember correctly, there were also issues around the fact that the Arcs are fields of the `WorkerState` struct, and borrowing one field for reading locked the whole `struct`, making it hard to modify the other field in the same context.

Reply:
I think there are techniques that we could use to deal with the borrowing, and ultimately, I think something like an `AtomicBool` may be simpler and easier to reason about.

That being said, we can probably tackle that later. I don't think we need to hold this feature up any more.